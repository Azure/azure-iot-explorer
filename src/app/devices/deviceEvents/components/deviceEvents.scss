/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
@import '../../../css/variables';
@import '../../../css/themes';

.device-events {
    .device-events-container {
        @include themify($themes) {
            border-top: 1px solid themed('contrastBorder');
        }
        margin-left: 15px;
        .device-events-content {
            @include themify($themes) {
                border-bottom: 1px solid themed('contrastBorder');
            }
            padding: 0 25px;
            width: 100%;
            pre {
                white-space: pre-wrap;
            }
        }
    }

    .consumer-group-label {
        font-weight: 600;
        font-size: 14px;
        @include themify($themes) {
            color: themed('textColor');
        }
    }

    .consumer-group-text-field {
        display: grid;
        padding-left: 24px;
        padding-bottom: 10px;
        @include themify($themes) {
            color: themed('textColor');
        }
        width: 85%;
    }

    .custom-event-hub-label {
        font-weight: 600;
        font-size: 14px;
        @include themify($themes) {
            color: themed('textColor');
        }
    }

    .pnp-telemetry {
        .column-value-text {
            .value-validation-error {
                @include themify($themes) {
                    color: themed('errorText');
                }
                padding-top: 10px;
                font-size: 14px;
            }
        }
    }

    .pnp-modeled-list {
        @include themify($themes) {
            background-color: themed('backgroundColor');
        }
        .list-header {
            font-weight: bold;
            border-bottom: 1px solid;
            @include themify($themes) {
                border-bottom-color: themed('borderColor');
            }
            padding-bottom: 5px;
        }
    }

    .item-summary {
        overflow: auto;
        width: 100%;
        border-bottom: 1px solid;
        @include themify($themes) {
            border-bottom-color: themed('borderColor');
        }
    }

    .stack-first-column {
        padding-left: 24px;
        min-width: 22%;
    }

    .stack-second-column {
        .custom-text-field {
            display: grid;
            padding-left: 24px;
            padding-bottom: 10px;
            @include themify($themes) {
                color: themed('textColor');
            }
            width: 80%;
        }
        width: 100%;
    }
}