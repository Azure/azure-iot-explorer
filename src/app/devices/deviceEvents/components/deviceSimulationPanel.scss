.ms-Panel-content {
    .cloudShellButton {
        cursor: pointer;
    }
    .collapsible-section {
        margin-top: 30px;
    }
}
