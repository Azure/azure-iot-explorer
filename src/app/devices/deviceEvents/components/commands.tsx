/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { useLocation, useHistory } from 'react-router-dom';
import { CommandBar, ICommandBarItemProps } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { CLEAR, CHECKED_CHECKBOX, EMPTY_CHECKBOX, START, STOP, NAVIGATE_BACK, REFRESH, REMOVE, CODE, UPLOAD } from '../../../constants/iconNames';
import { getComponentNameFromQueryString } from '../../../shared/utils/queryStringHelper';
import { usePnpStateContext } from '../../pnp/context/pnpStateContext';
import { getBackUrl } from '../../pnp/utils';
import { AppInsightsClient } from '../../../shared/appTelemetry/appInsightsClient';
import { TELEMETRY_USER_ACTIONS } from '../../../../app/constants/telemetry';
import { useDeviceEventsStateContext } from '../context/deviceEventsStateContext';
import './deviceEvents.scss';

export interface CommandsProps {
    startDisabled: boolean;
    monitoringData: boolean;
    showPnpModeledEvents: boolean;
    showSimulationPanel: boolean;
    showContentTypePanel: boolean;
    setMonitoringData: (monitoringData: boolean) => void;
    setShowPnpModeledEvents: (showPnpModeledEvents: boolean) => void;
    setShowSimulationPanel: (showSimulationPanel: boolean) => void;
    setShowContentTypePanel: (showDecoderPanel: boolean) => void;
    fetchData(): void;
    stopFetching(): void;
}

export const Commands: React.FC<CommandsProps> = ({
    startDisabled,
    monitoringData,
    showPnpModeledEvents,
    showSimulationPanel,
    showContentTypePanel,
    setMonitoringData,
    setShowPnpModeledEvents,
    setShowSimulationPanel,
    setShowContentTypePanel,
    fetchData,
    stopFetching}) => {

    const {t} = useTranslation();
    const { search, pathname } = useLocation();
    const history = useHistory();
    const { getModelDefinition } = usePnpStateContext();
    const componentName = getComponentNameFromQueryString(search); // if component name exist, we are in pnp context
    const [ state, api ] = useDeviceEventsStateContext();

    const createCommandBarItems = (): ICommandBarItemProps[] => {
        if (componentName) {
            return [createStartMonitoringCommandItem(),
                createPnpModeledEventsCommandItem(),
                createRefreshCommandItem(),
                createClearCommandItem()
            ];
        }
        else {
            return [createStartMonitoringCommandItem(),
                createClearCommandItem(),
                createSimulationCommandItem(),
                createContentTypeCommandItem()
            ];
        }
    };

    const createClearCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: t(ResourceKeys.deviceEvents.command.clearEvents),
            iconProps: {
                iconName: REMOVE
            },
            key: CLEAR,
            name: t(ResourceKeys.deviceEvents.command.clearEvents),
            onClick: onClearData
        };
    };

    const createPnpModeledEventsCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: 'Show modeled events',
            iconProps: {
                iconName: showPnpModeledEvents ? CHECKED_CHECKBOX : EMPTY_CHECKBOX
            },
            key: EMPTY_CHECKBOX,
            name: 'Show modeled events',
            onClick: onShowPnpModeledEvents
        };
    };

    // tslint:disable-next-line: cyclomatic-complexity
    const createStartMonitoringCommandItem = (): ICommandBarItemProps => {
        const label = monitoringData ? t(ResourceKeys.deviceEvents.command.stop) : t(ResourceKeys.deviceEvents.command.start);
        const icon = monitoringData ? STOP : START;
        return {
            ariaLabel: label,
            disabled: startDisabled,
            iconProps: {
                iconName: icon
            },
            key: icon,
            name: label,
            onClick: onToggleStart
        };
    };

    const createSimulationCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: t(ResourceKeys.deviceEvents.command.simulate),
            iconProps: {
                iconName: CODE
            },
            key: t(ResourceKeys.deviceEvents.command.simulate),
            name: t(ResourceKeys.deviceEvents.command.simulate),
            onClick: onToggleSimulationPanel
        };
    };

    const createRefreshCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: t(ResourceKeys.deviceEvents.command.refresh),
            disabled: state.formMode === 'updating',
            iconProps: {iconName: REFRESH},
            key: REFRESH,
            name: t(ResourceKeys.deviceEvents.command.refresh),
            onClick: getModelDefinition
        };
    };

    const createNavigateBackCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: t(ResourceKeys.deviceEvents.command.close),
            iconProps: {iconName: NAVIGATE_BACK},
            key: NAVIGATE_BACK,
            name: t(ResourceKeys.deviceEvents.command.close),
            onClick: handleClose
        };
    };

    const createContentTypeCommandItem = (): ICommandBarItemProps => {
        return {
            ariaLabel: t(ResourceKeys.deviceEvents.command.customizeContentType),
            disabled: state.formMode !== 'upserted' && state.formMode !== 'initialized'
                        && state.formMode !== 'setDecoderSucceeded' && state.formMode !== 'setDecoderFailed',
            iconProps: {
                iconName: UPLOAD
            },
            key: UPLOAD,
            name: t(ResourceKeys.deviceEvents.command.customizeContentType),
            onClick: onToggleContentTypePanel
        };
    };

    const handleClose = () => {
        const path = pathname.replace(/\/ioTPlugAndPlayDetail\/events\/.*/, ``);
        history.push(getBackUrl(path, search));
    };

    const onToggleStart = () => {
        if (monitoringData) {
            setMonitoringData(false);
            stopFetching();
        } else {
            fetchData();
            setMonitoringData(true);
            if (componentName) {
                AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.PNP_START_TELEMETRY);
            } else {
                AppInsightsClient.trackUserAction(TELEMETRY_USER_ACTIONS.START_DEVICE_TELEMETRY);
            }
        }
    };

    const onClearData = () => {
        api.clearEventsMonitoring();
    };

    const onShowPnpModeledEvents = () => {
        setShowPnpModeledEvents(!showPnpModeledEvents);
    };

    const onToggleSimulationPanel = () => {
        setShowSimulationPanel(!showSimulationPanel);
    };

    const onToggleContentTypePanel = () => {
        setShowContentTypePanel(!showContentTypePanel);
    };

    return (
        <>
            <CommandBar
                items={createCommandBarItems()}
                farItems={componentName && [createNavigateBackCommandItem()]}
            />
        </>
    );
};
