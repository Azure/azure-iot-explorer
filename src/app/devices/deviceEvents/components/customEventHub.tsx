/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { Toggle, TextField, Stack } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { isValidEventHubConnectionString } from '../../../shared/utils/hubConnectionStringHelper';
import './deviceEvents.scss';

export interface CustomEventHubProps {
    monitoringData: boolean;
    useBuiltInEventHub: boolean;
    customEventHubConnectionString: string;
    setUseBuiltInEventHub: (monitoringData: boolean) => void;
    setCustomEventHubConnectionString: (customEventHubConnectionString: string) => void;
    setHasError: (hasError: boolean) => void;
}

export const CustomEventHub: React.FC<CustomEventHubProps> = ({
    monitoringData,
    useBuiltInEventHub,
    customEventHubConnectionString,
    setUseBuiltInEventHub,
    setCustomEventHubConnectionString,
    setHasError}) => {

    const { t } = useTranslation();
    const [ error, setError ] = React.useState<string>();

    React.useEffect(() => {
        if (!isValidEventHubConnectionString(customEventHubConnectionString)) {
            setError(t(ResourceKeys.deviceEvents.customEventHub.connectionString.error));
            setHasError(true);
        }
        else {
            setError('');
            setHasError(false);
        }
    },              [customEventHubConnectionString]);

    const toggleChange = () => {
        setUseBuiltInEventHub(!useBuiltInEventHub);
    };

    const customEventHubConnectionStringChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        setCustomEventHubConnectionString(newValue);
    };

    return (
        <Stack horizontal={true} horizontalAlign="space-between">
            <Toggle
                className="stack-first-column"
                checked={useBuiltInEventHub}
                ariaLabel={t(ResourceKeys.deviceEvents.toggleUseDefaultEventHub.label)}
                label={t(ResourceKeys.deviceEvents.toggleUseDefaultEventHub.label)}
                onText={t(ResourceKeys.deviceEvents.toggleUseDefaultEventHub.on)}
                offText={t(ResourceKeys.deviceEvents.toggleUseDefaultEventHub.off)}
                onChange={toggleChange}
                disabled={monitoringData}
            />
            {!useBuiltInEventHub &&
                <Stack horizontal={false} className="stack-second-column">
                    <TextField
                        className="custom-text-field"
                        label={t(ResourceKeys.deviceEvents.customEventHub.connectionString.label)}
                        ariaLabel={t(ResourceKeys.deviceEvents.customEventHub.connectionString.label)}
                        underlined={true}
                        value={customEventHubConnectionString}
                        disabled={monitoringData}
                        onChange={customEventHubConnectionStringChange}
                        placeholder={t(ResourceKeys.deviceEvents.customEventHub.connectionString.placeHolder)}
                        errorMessage={error}
                        required={true}
                    />
                </Stack>
            }
        </Stack>
    );
};
