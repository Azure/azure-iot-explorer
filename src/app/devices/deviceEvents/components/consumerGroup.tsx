/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { ITextFieldProps, TextField } from '@fluentui/react';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { LabelWithTooltip } from '../../../shared/components/labelWithTooltip';
import './deviceEvents.scss';

export interface ConsumerGroupProps {
    monitoringData: boolean;
    consumerGroup: string;
    setConsumerGroup: (consumerGroup: string) => void;
}

export const ConsumerGroup: React.FC<ConsumerGroupProps> = ({monitoringData, consumerGroup, setConsumerGroup}) => {

    const { t } = useTranslation();
    const renderConsumerGroupLabel = (textFieldProps: ITextFieldProps) => (
        <LabelWithTooltip
            className={'consumer-group-label'}
            tooltipText={t(ResourceKeys.deviceEvents.consumerGroups.tooltip)}
        >
            {textFieldProps.label}
        </LabelWithTooltip>
    );

    const consumerGroupChange = (event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>, newValue?: string) => {
        setConsumerGroup(newValue);
    };

    return (
        <TextField
            className={'consumer-group-text-field'}
            onRenderLabel={renderConsumerGroupLabel}
            label={t(ResourceKeys.deviceEvents.consumerGroups.label)}
            ariaLabel={t(ResourceKeys.deviceEvents.consumerGroups.label)}
            underlined={true}
            value={consumerGroup}
            disabled={monitoringData}
            onChange={consumerGroupChange}
        />
    );
};
