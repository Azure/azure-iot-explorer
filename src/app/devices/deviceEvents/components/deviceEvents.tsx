/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { Stack } from '@fluentui/react';
import { useLocation } from 'react-router-dom';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../../shared/utils/queryStringHelper';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import { MonitorEventsParameters } from '../../../api/parameters/deviceParameters';
import { DEFAULT_CONSUMER_GROUP, WEBSOCKET_ENDPOINT } from '../../../constants/apiConstants';
import { HeaderView } from '../../../shared/components/headerView';
import { useDeviceEventsStateContext } from '../context/deviceEventsStateContext';
import { usePnpStateContext } from '../../pnp/context/pnpStateContext';
import { MultiLineShimmer } from '../../../shared/components/multiLineShimmer';
import { DeviceSimulationPanel } from './deviceSimulationPanel';
import { Commands } from './commands';
import { CustomEventHub } from './customEventHub';
import { ConsumerGroup } from './consumerGroup';
import { StartTime } from './startTime';
import { DeviceContentTypePanel } from './deviceContentTypePanel';
import { Loader } from './loader';
import { EventsContent } from './eventsContent';
import { SystemPropertyCheckBox } from './systemPropertyCheckBox';
import './deviceEvents.scss';

let client: WebSocket;
export const DeviceEvents: React.FC = () => {
    const { search } = useLocation();
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    const [ state, api ] = useDeviceEventsStateContext();
    const decoderPrototype = state.contentType.decoderPrototype;

    // event hub settings
    const [consumerGroup, setConsumerGroup] = React.useState(DEFAULT_CONSUMER_GROUP);
    const [specifyStartTime, setSpecifyStartTime] = React.useState<boolean>(false);
    const [startTime, setStartTime] = React.useState<Date>();
    const [useBuiltInEventHub, setUseBuiltInEventHub] = React.useState<boolean>(true);
    const [customEventHubConnectionString, setCustomEventHubConnectionString] = React.useState<string>(undefined);
    const [showSystemProperties, setShowSystemProperties] = React.useState<boolean>(false);

    // event message state
    const [monitoringData, setMonitoringData] = React.useState<boolean>(false);
    const [startDisabled, setStartDisabled] = React.useState<boolean>(false);
    const [hasError, setHasError] = React.useState<boolean>(false);

    // pnp events specific
    const { pnpState, } = usePnpStateContext();
    const isLoading = pnpState.modelDefinitionWithSource.synchronizationStatus === SynchronizationStatus.working;
    const [showPnpModeledEvents, setShowPnpModeledEvents] = React.useState(false);

    // simulation specific
    const [showSimulationPanel, setShowSimulationPanel] = React.useState(false);

    // message content type specific
    const [showContentTypePanel, setShowContentTypePanel] = React.useState(false);

    React.useEffect(
        () => {
            return () => {
                stopMonitoring();
                client?.close();
            };
        },
        []);

    React.useEffect(
        () => {
            client = new WebSocket(WEBSOCKET_ENDPOINT);
        },
        []);

    React.useEffect(
        () => {
            if (monitoringData) {
                client.onmessage = message => {
                    api.setEvents(JSON.parse(message.data));
                };
            }
        },
        [monitoringData]);

    React.useEffect(    // tslint:disable-next-line: cyclomatic-complexity
        () => {
            if (state.formMode === 'updating' ||
                // when specifying start time, valid time need to be provided
                (specifyStartTime && (!startTime || hasError)) ||
                // when using custom event hub, both valid connection string and name need to be provided
                (!useBuiltInEventHub && (!customEventHubConnectionString || hasError))) {
                setStartDisabled(true);
            }
            else {
                setStartDisabled(false);
            }
        },
        [hasError, state.formMode, useBuiltInEventHub, customEventHubConnectionString, specifyStartTime, startTime]);

    const onSystemPropertyCheckBoxChange = (ev?: React.FormEvent<HTMLElement | HTMLInputElement>, checked?: boolean) => {
        setShowSystemProperties(!!checked);
    };

    const renderCommands = () => {
        return (
            <Commands
                startDisabled={startDisabled}
                monitoringData={monitoringData}
                showPnpModeledEvents={showPnpModeledEvents}
                showSimulationPanel={showSimulationPanel}
                showContentTypePanel={showContentTypePanel}
                setMonitoringData={setMonitoringData}
                setShowPnpModeledEvents={setShowPnpModeledEvents}
                setShowSimulationPanel={setShowSimulationPanel}
                setShowContentTypePanel={setShowContentTypePanel}
                fetchData={fetchData}
                stopFetching={stopMonitoring}
            />
        );
    };

    const renderConsumerGroup = () => {
        return (
            <div className="horizontal-item">
                <ConsumerGroup
                    monitoringData={monitoringData}
                    consumerGroup={consumerGroup}
                    setConsumerGroup={setConsumerGroup}
                />
            </div>
        );
    };

    const renderStartTimePicker = () => {
        return (
            <StartTime
                monitoringData={monitoringData}
                specifyStartTime={specifyStartTime}
                startTime={startTime}
                setSpecifyStartTime={setSpecifyStartTime}
                setStartTime={setStartTime}
                setHasError={setHasError}
            />
        );
    };

    const renderCustomEventHub = () => {
        return (
            <div className="horizontal-item">
                <CustomEventHub
                    monitoringData={monitoringData}
                    useBuiltInEventHub={useBuiltInEventHub}
                    customEventHubConnectionString={customEventHubConnectionString}
                    setUseBuiltInEventHub={setUseBuiltInEventHub}
                    setCustomEventHubConnectionString={setCustomEventHubConnectionString}
                    setHasError={setHasError}
                />
            </div>
        );
    };

    const stopMonitoring = () => {
        api.stopEventsMonitoring();
    };

    const fetchData = () => {
        client.onopen = () => { // intentionally blank
        };
        let parameters: MonitorEventsParameters = {
            consumerGroup,
            decoderPrototype,
            deviceId,
            moduleId,
            startTime
        };

        if (!useBuiltInEventHub) {
            parameters = {
                ...parameters,
                customEventHubConnectionString
            };
        }
        api.startEventsMonitoring(parameters);
    };

    const onToggleSimulationPanel = () => {
        setShowSimulationPanel(!showSimulationPanel);
    };

    const onToggleContentTypePanel = () => {
        setShowContentTypePanel(!showContentTypePanel);
    };

    if (isLoading) {
        return <MultiLineShimmer />;
    }

    return (
        <Stack className="device-events" key="device-events">
            {renderCommands()}
            <HeaderView
                headerText={ResourceKeys.deviceEvents.headerText}
                tooltip={ResourceKeys.deviceEvents.tooltip}
            />
            {renderConsumerGroup()}
            {renderStartTimePicker()}
            {renderCustomEventHub()}
            <DeviceSimulationPanel
                showSimulationPanel={showSimulationPanel}
                onToggleSimulationPanel={onToggleSimulationPanel}
            />
            <DeviceContentTypePanel
                showContentTypePanel={showContentTypePanel}
                onToggleContentTypePanel={onToggleContentTypePanel}
            />
            <div className="device-events-container">
                <SystemPropertyCheckBox
                     showSystemProperties={showSystemProperties}
                     showPnpModeledEvents={showPnpModeledEvents}
                     setShowSystemProperties={onSystemPropertyCheckBoxChange}
                />
                <Loader monitoringData={monitoringData}/>
                <EventsContent showPnpModeledEvents={showPnpModeledEvents} showSystemProperties={showSystemProperties}/>
            </div>
        </Stack>
    );
};
