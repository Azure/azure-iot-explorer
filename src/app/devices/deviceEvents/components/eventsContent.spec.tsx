/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { REPOSITORY_LOCATION_TYPE } from '../../../constants/repositoryLocationTypes';
import { ResourceKeys } from '../../../../localization/resourceKeys';
import { SynchronizationStatus } from '../../../api/models/synchronizationStatus';
import * as pnpStateContext from '../../pnp/context/pnpStateContext';
import { ErrorBoundary } from '../../shared/components/errorBoundary';
import { pnpStateInitial, PnpStateInterface } from '../../pnp/state';
import { testModelDefinition } from '../../pnp/components/deviceEvents/testData';
import * as deviceEventsStateContext from '../context/deviceEventsStateContext';
import { getInitialDeviceEventsState } from '../state';
import { EventsContent } from './eventsContent';

const pathname = `#/devices/detail/events/?id=device1`;
jest.mock('react-router-dom', () => ({
    useHistory: () => ({ push: jest.fn() }),
    useLocation: () => ({ search: `?deviceId=device1`, pathname, push: jest.fn() })
}));

describe('EventsContent', () => {
    describe('EventsContent in non-pnp context', () => {
        const events = [{
            body: {
                humid: 123
            },
            enqueuedTime: 'Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)',
            properties: {
            'iothub-message-schema': 'humid'
            }
        }];

        it('matches snapshot', () => {
            const startEventsMonitoring = jest.fn();
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps(), startEventsMonitoring}]);
            expect(shallow(<EventsContent showPnpModeledEvents={false} showSystemProperties={false}/>)).toMatchSnapshot();
        });

        it('renders events', () => {
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps()}]);
            jest.spyOn(pnpStateContext, 'usePnpStateContext').mockReturnValue({ pnpState: pnpStateInitial(), dispatch: jest.fn()});
            const wrapper = mount(<EventsContent showPnpModeledEvents={false} showSystemProperties={false}/>);
            const rawTelemetry = wrapper.find('article');
            expect(rawTelemetry).toHaveLength(1);
        });
    });

    describe('EventsContent in pnp context', () => {
        const getModelDefinitionMock = jest.fn();

        const mockFetchedState = () => {
            const pnpState: PnpStateInterface = {
                ...pnpStateInitial(),
                modelDefinitionWithSource: {
                    payload: {
                        isModelValid: true,
                        modelDefinition: testModelDefinition,
                        source: REPOSITORY_LOCATION_TYPE.Public,
                    },
                    synchronizationStatus: SynchronizationStatus.fetched
                }
            };
            jest.spyOn(pnpStateContext, 'usePnpStateContext').mockReturnValue({pnpState, dispatch: jest.fn(), getModelDefinition: getModelDefinitionMock});
        };

        it('matches snapshot while interface definition is retrieved', () => {
            mockFetchedState();
            expect(shallow(<EventsContent showPnpModeledEvents={true} showSystemProperties={false}/>)).toMatchSnapshot();
        });

        it('renders events which body\'s value type is wrong with expected columns', () => {
            const events = [{
                body: {
                    humid: '123' // intentionally set a value which type is double
                },
                enqueuedTime: 'Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)',
                systemProperties: {
                'iothub-message-schema': 'humid'
                }
            }];
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events, formMode: 'fetched'},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps()}]);
            mockFetchedState();

            const wrapper = mount(<EventsContent showPnpModeledEvents={true} showSystemProperties={false}/>);
            const errorBoundary = wrapper.find(ErrorBoundary);
            expect(errorBoundary.children().at(1).props().children.props.children).toEqual('humid (Temperature)');
            expect(errorBoundary.children().at(2).props().children.props.children).toEqual('double'); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[0]).toEqual(JSON.stringify(events[0].body, undefined, 2)); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[1].props['aria-label']).toEqual(ResourceKeys.deviceEvents.columns.validation.value.label); // tslint:disable-line:no-magic-numbers
        });

        it('renders events which body\'s key name is wrong with expected columns', () => {
            const events = [{
                body: {
                    'non-matching-key': 0
                },
                enqueuedTime: 'Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)',
                systemProperties: {
                'iothub-message-schema': 'humid'
                }
            }];
            mockFetchedState();
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events, formMode: 'fetched'},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps()}]);
            mockFetchedState();

            const wrapper = mount(<EventsContent showPnpModeledEvents={true} showSystemProperties={false}/>);
            const errorBoundary = wrapper.find(ErrorBoundary);
            expect(errorBoundary.children().at(1).props().children.props.children).toEqual('humid (Temperature)');
            expect(errorBoundary.children().at(2).props().children.props.children).toEqual('double'); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[0]).toEqual(JSON.stringify(events[0].body, undefined, 2)); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[1].props.className).toEqual('value-validation-error'); // tslint:disable-line:no-magic-numbers
        });

        it('renders events when body is exploded and schema is not provided in system properties', () => {
            const events = [{
                body: {
                    'humid': 0,
                    'humid-foo': 'test'
                },
                enqueuedTime: 'Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)',
                systemProperties: {}
            }];
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events, formMode: 'fetched'},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps()}]);
            mockFetchedState();

            const wrapper = shallow(<EventsContent showPnpModeledEvents={true} showSystemProperties={false}/>);
            let errorBoundary = wrapper.find(ErrorBoundary).first();
            expect(errorBoundary.children().at(1).props().children.props.children).toEqual('humid (Temperature)');
            expect(errorBoundary.children().at(2).props().children.props.children).toEqual('double'); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[0]).toEqual(JSON.stringify({humid: 0}, undefined, 2)); // tslint:disable-line:no-magic-numbers

            errorBoundary = wrapper.find(ErrorBoundary).at(1);
            expect(errorBoundary.children().at(1).props().children.props.children).toEqual('--');
            expect(errorBoundary.children().at(2).props().children.props.children).toEqual('--'); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[0]).toEqual(JSON.stringify({'humid-foo': 'test'}, undefined, 2)); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children[1].props.className).toEqual('value-validation-error'); // tslint:disable-line:no-magic-numbers
        });

        it('renders events which body\'s key name is not populated', () => {
            const events = [{
                body: {
                    'non-matching-key': 0
                },
                enqueuedTime: 'Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)'
            }];
            jest.spyOn(deviceEventsStateContext, 'useDeviceEventsStateContext').mockReturnValue(
                [{...getInitialDeviceEventsState(), message: events, formMode: 'fetched'},
                    {...deviceEventsStateContext.getInitialDeviceEventsOps()}]);
            mockFetchedState();

            const wrapper = shallow(<EventsContent showPnpModeledEvents={true} showSystemProperties={false}/>);
            const errorBoundary = wrapper.find(ErrorBoundary);
            expect(errorBoundary.children().at(1).props().children.props.children).toEqual('--');
            expect(errorBoundary.children().at(2).props().children.props.children).toEqual('--'); // tslint:disable-line:no-magic-numbers
            expect(errorBoundary.children().at(4).props().children.props.children).toEqual(JSON.stringify(events[0].body, undefined, 2)); // tslint:disable-line:no-magic-numbers
        });
    });
});