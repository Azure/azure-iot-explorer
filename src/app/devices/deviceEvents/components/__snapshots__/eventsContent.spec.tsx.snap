// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsContent EventsContent in non-pnp context matches snapshot 1`] = `
<div
  className=""
>
  <article
    className="device-events-content"
    key="0"
  >
    <h5>
      Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)
      :
    </h5>
    <pre>
      {
  "body": {
    "humid": 123
  },
  "enqueuedTime": "Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)",
  "properties": {
    "iothub-message-schema": "humid"
  }
}
    </pre>
  </article>
</div>
`;

exports[`EventsContent EventsContent in pnp context matches snapshot while interface definition is retrieved 1`] = `
<div
  className=""
>
  <div
    className="pnp-modeled-list"
  >
    <div
      className="list-header list-header-uncollapsible flex-grid-row"
    >
      <span
        className="col-sm2"
      >
        deviceEvents.columns.timestamp
      </span>
      <span
        className="col-sm2"
      >
        deviceEvents.columns.displayName
      </span>
      <span
        className="col-sm2"
      >
        deviceEvents.columns.schema
      </span>
      <span
        className="col-sm2"
      >
        deviceEvents.columns.unit
      </span>
      <span
        className="col-sm4"
      >
        deviceEvents.columns.value
      </span>
    </div>
  </div>
  <section
    role="feed"
  >
    <article
      className="list-item event-list-item"
      key="0"
      role="article"
    >
      <section
        className="flex-grid-row item-summary"
      >
        <ErrorBoundary
          error="errorBoundary.text"
        >
          <div
            className="col-sm2"
          >
            <StyledLabelBase
              aria-label="deviceEvents.columns.timestamp"
            >
              Wed Feb 17 2021 16:06:00 GMT-0800 (Pacific Standard Time)
            </StyledLabelBase>
          </div>
          <div
            className="col-sm2"
          >
            <StyledLabelBase
              aria-label="deviceEvents.columns.displayName"
            >
              --
            </StyledLabelBase>
          </div>
          <div
            className="col-sm2"
          >
            <StyledLabelBase
              aria-label="deviceEvents.columns.schema"
            >
              --
            </StyledLabelBase>
          </div>
          <div
            className="col-sm2"
          >
            <StyledLabelBase
              aria-label="deviceEvents.columns.unit"
            >
              <SemanticUnit />
            </StyledLabelBase>
          </div>
          <div
            className="column-value-text col-sm4"
          >
            <StyledLabelBase
              aria-label="deviceEvents.columns.value"
            >
              {
  "humid": 123
}
            </StyledLabelBase>
          </div>
        </ErrorBoundary>
      </section>
    </article>
  </section>
</div>
`;
